src/source-control/GitSourceControl.ts

```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
+     */
+    const commitChildrenMap = new Map<string, Set<string>>();
+    await Promise.all(
+      refs.map((ref) => {
+        return new Promise(async (resolve, _) => {
+          const history = (await repo.getBranchCommit(ref)).history();
+          let stopped = false;
+
+          history.on("commit", (nodegitCommit: nodegit.Commit) => {
+            if (stopped) {
+              return;
+            }
+            const sha = nodegitCommit.sha();
+
+            // If the commit is already in `commitHashDict`, stop because we
+            // have already processed it.
+            if (hashMap.has(sha)) {
+              stopped = true;
+              return;
+            }
+
+            const commitSignature: CommitSignature = {
+              name: nodegitCommit.author().name(),
+              email: nodegitCommit.author().email(),
+            };
+
+            // Create a new commit and add it to `commitHashDict`.
+            const newCommit: Commit = {
+              title: nodegitCommit.summary(),
+              hash: sha,
+              timestamp: nodegitCommit.date(),
+              author: commitSignature,
+              committer: commitSignature,
+              refNames: [],
+              parentCommits: [],
+              childCommits: [],
+            };
+            // Use Immer?
+            hashMap.set(sha, newCommit);
+
+            // Update the commit's children.
+            const parentHashes = nodegitCommit
+              .parents()
+              .map((parent) => parent.tostrS());
+            parentHashes.forEach((parentHash) => {
+              if (!commitChildrenMap.has(parentHash)) {
+                commitChildrenMap.set(parentHash, new Set());
+              }
+              commitChildrenMap.get(parentHash)!.add(sha);
+            });
+
+            // Stop if this is the common ancestor commit
+            if (nodegitCommit.id().equal(commonAncestorCommitOid)) {
+              stopped = true;
+            }
+          });
+
+          history.on("end", () => resolve());
+          history.start();
+        });
+      }),
+    );
+
+    // Build relationships
+    commitChildrenMap.forEach((childrenHashes, parentHash) => {
+      const parentCommit = hashMap.get(parentHash);
+      if (!parentCommit) {
+        return;
+      }
+      childrenHashes.forEach((childHash) => {
+        const childCommit: Commit = hashMap.get(childHash)!;
+        childCommit.parentCommits = Array.from(
+          new Set([...childCommit.parentCommits, parentHash]),
+        );
+        parentCommit.childCommits = Array.from(
+          new Set([...parentCommit.childCommits, childHash]),
+        );
+        hashMap.set(childHash, childCommit);
+      });
+      hashMap.set(parentHash, parentCommit);
+    });
+
+    // Inject branch names
+
+    refs.forEach((ref) => {
+      const branchName = ref.name();
+      const branchSha = ref.target().tostrS();
+      hashMap.get(branchSha)?.refNames.push(branchName);
```

Comment:
Probably a relic from the old code, but `branchName` isn't accurate
```suggestion
      const refName = ref.name();
      const branchSha = ref.target().tostrS();
      hashMap.get(branchSha)?.refNames.push(refName);
```

Reply:
👍